File: sdk/clojure/src/hitechcloud/sdk/cloud_service.clj
(ns hitechcloud.sdk.cloud_service
  (:require [hitechcloud.sdk.http-client :as http]))

;; Cloud Service: VM shutdown, reset, hostname, IP pools, networks, interfaces

(defn shutdown_vm "POST /api/service/{id}/vms/{vmid}/shutdown - Perform graceful shutdown" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/shutdown" params))

(defn reset_vm "POST /api/service/{id}/vms/{vmid}/reset - Reset virtual server power" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/reset" params))

(defn change_hostname "POST /api/service/{id}/vms/{vmid}/hostname - Update VM hostname" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/hostname" params))

(defn list_ip_pools "GET /api/service/{id}/vms/{vmid}/ippool - List IP pools" [client & [params]] (http-get client "/api/service/{id}/vms/{vmid}/ippool" params))

(defn allocate_ip "POST /api/service/{id}/vms/{vmid}/ippool/{pool} - Allocate new IP" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/ippool/{pool}" params))

(defn list_networks "GET /api/service/{id}/networks - List available networks" [client & [params]] (http-get client "/api/service/{id}/networks" params))

(defn add_interface "POST /api/service/{id}/vms/{vmid}/interfaces - Add network interface" [client & [params]] (http-post client "/api/service/{id}/vms/{vmid}/interfaces" params))

(defn get_interface "GET /api/service/{id}/vms/{vmid}/interfaces/{iface} - Get interface details" [client & [params]] (http-get client "/api/service/{id}/vms/{vmid}/interfaces/{iface}" params))

(defn update_interface "PUT /api/service/{id}/vms/{vmid}/interfaces/{iface} - Update interface" [client & [params]] (http-put client "/api/service/{id}/vms/{vmid}/interfaces/{iface}" params))

(defn remove_interface "DELETE /api/service/{id}/vms/{vmid}/interfaces/{iface} - Remove interface" [client & [params]] (http-delete client "/api/service/{id}/vms/{vmid}/interfaces/{iface}" params))

File: sdk/fsharp/src/HiTechCloudSDK/CloudService.fs
namespace HiTechCloud.SDK

/// Cloud Service: VM shutdown, reset, hostname, IP pools, networks, interfaces
type CloudServiceResource(httpClient: HttpClient) =

    /// POST /api/service/{id}/vms/{vmid}/shutdown - Perform graceful shutdown
    member this.ShutdownVm() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/shutdown")
        }

    /// POST /api/service/{id}/vms/{vmid}/reset - Reset virtual server power
    member this.ResetVm() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/reset")
        }

    /// POST /api/service/{id}/vms/{vmid}/hostname - Update VM hostname
    member this.ChangeHostname() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/hostname")
        }

    /// GET /api/service/{id}/vms/{vmid}/ippool - List IP pools
    member this.ListIpPools() =
        async {
            return! httpClient.getAsync("/api/service/{id}/vms/{vmid}/ippool")
        }

    /// POST /api/service/{id}/vms/{vmid}/ippool/{pool} - Allocate new IP
    member this.AllocateIp() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/ippool/{pool}")
        }

    /// GET /api/service/{id}/networks - List available networks
    member this.ListNetworks() =
        async {
            return! httpClient.getAsync("/api/service/{id}/networks")
        }

    /// POST /api/service/{id}/vms/{vmid}/interfaces - Add network interface
    member this.AddInterface() =
        async {
            return! httpClient.postAsync("/api/service/{id}/vms/{vmid}/interfaces")
        }

    /// GET /api/service/{id}/vms/{vmid}/interfaces/{iface} - Get interface details
    member this.GetInterface() =
        async {
            return! httpClient.getAsync("/api/service/{id}/vms/{vmid}/interfaces/{iface}")
        }

    /// PUT /api/service/{id}/vms/{vmid}/interfaces/{iface} - Update interface
    member this.UpdateInterface() =
        async {
            return! httpClient.putAsync("/api/service/{id}/vms/{vmid}/interfaces/{iface}")
        }

    /// DELETE /api/service/{id}/vms/{vmid}/interfaces/{iface} - Remove interface
    member this.RemoveInterface() =
        async {
            return! httpClient.deleteAsync("/api/service/{id}/vms/{vmid}/interfaces/{iface}")
        }

